From: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
Date: Wed, 24 Sep 2025 12:04:02 +0800
Subject: [PATCH] net: ethernet: mtk_ppe: change to internal QoS mode (PPPQ)

Adapted for mainline OpenWrt (stock mtk_ppe): adds the qos_toggle field and
the per-port-per-queue output-device path. Drops the SDK proprietary-debugfs
hunks (mtk_ppe_internal_debugfs_*, not present in mainline); qos_toggle is set
at probe by 999-ppe-36. Adds MTK_QDMA_QUEUE_MASK locally (was in the dropped
eth-26). Keeps the flow_offload->ct plumbing (needed by 999-ppe-11).

Signed-off-by: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
---
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.h
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
@@ -26,6 +26,7 @@
 #define MTK_DSA_PORT_MASK	GENMASK(2, 0)
 
 #define MTK_QDMA_NUM_QUEUES	16
+#define MTK_QDMA_QUEUE_MASK	((1ULL << MTK_QDMA_NUM_QUEUES) - 1)
 #define MTK_QDMA_PAGE_SIZE	2048
 #define MTK_MAX_RX_LENGTH	1536
 #define MTK_MAX_RX_LENGTH_2K	2048
@@ -1290,6 +1291,7 @@
 
 	struct metadata_dst		*dsa_meta[MTK_MAX_DSA_PORTS];
 
+	u8				qos_toggle;
 	struct mtk_ppe			*ppe[3];
 	struct rhashtable		flow_table;
 
@@ -1426,6 +1428,18 @@
 	return MTK_FOE_IB2_MULTICAST;
 }
 
+static inline int
+mtk_ppe_check_pppq_path(struct mtk_mac *mac, struct net_device *idev, int dsa_port)
+{
+	bool wifi_rx = !!(idev->ieee80211_ptr);
+
+	if ((dsa_port >= 0 && dsa_port <= 4) ||
+	    (dsa_port == 5 && wifi_rx))
+		return 1;
+
+	return 0;
+}
+
 /* read the hardware status register */
 void mtk_stats_update_mac(struct mtk_mac *mac);
 
--- a/drivers/net/ethernet/mediatek/mtk_ppe_offload.c
+++ b/drivers/net/ethernet/mediatek/mtk_ppe_offload.c
@@ -9,6 +9,8 @@
 #include <linux/ipv6.h>
 #include <net/flow_offload.h>
 #include <net/pkt_cls.h>
+#include <net/netfilter/nf_conntrack_acct.h>
+#include <net/netfilter/nf_flow_table.h>
 #include <net/dsa.h>
 #include "mtk_eth_soc.h"
 #include "mtk_wed.h"
@@ -189,13 +191,17 @@
 
 static int
 mtk_flow_set_output_device(struct mtk_eth *eth, struct mtk_foe_entry *foe,
-			   struct net_device *dev, const u8 *dest_mac,
+			   struct net_device *idev, struct net_device *odev,
+			   struct flow_cls_offload *f, const u8 *dest_mac,
 			   int *wed_index)
 {
 	struct mtk_wdma_info info = {};
+	struct nf_conn *ct = NULL;
+	struct mtk_mac *mac;
+	u32 ct_mark = 0;
 	int pse_port, dsa_port, queue;
 
-	if (mtk_flow_get_wdma_info(dev, dest_mac, &info) == 0) {
+	if (mtk_flow_get_wdma_info(odev, dest_mac, &info) == 0) {
 		mtk_foe_entry_set_wdma(eth, foe, info.wdma_idx, info.queue,
 				       info.bss, info.wcid, info.amsdu);
 		if (mtk_is_netsys_v2_or_greater(eth)) {
@@ -219,24 +225,52 @@
 		goto out;
 	}
 
-	dsa_port = mtk_flow_get_dsa_port(&dev);
+	dsa_port = mtk_flow_get_dsa_port(&odev);
 
-	if (dev == eth->netdev[0])
+	if (odev == eth->netdev[0]) {
+		mac = eth->mac[0];
 		pse_port = PSE_GDM1_PORT;
-	else if (dev == eth->netdev[1])
+	} else if (odev == eth->netdev[1]) {
+		mac = eth->mac[1];
 		pse_port = PSE_GDM2_PORT;
-	else if (dev == eth->netdev[2])
+	} else if (odev == eth->netdev[2]) {
+		mac = eth->mac[2];
 		pse_port = PSE_GDM3_PORT;
-	else
+	} else
 		return -EOPNOTSUPP;
 
 	if (dsa_port >= 0) {
 		mtk_foe_entry_set_dsa(eth, foe, dsa_port);
 		queue = 3 + dsa_port;
 	} else {
-		queue = pse_port - 1;
+		queue = (pse_port == PSE_GDM3_PORT) ? 2 : pse_port - 1;
+	}
+
+	if (f->flow && f->flow->ct) {
+		ct = f->flow->ct;
+#if defined(CONFIG_NF_CONNTRACK_MARK)
+		ct_mark = ct->mark;
+#endif
+	}
+
+	if (eth->qos_toggle == 2 && mtk_ppe_check_pppq_path(mac, idev, dsa_port))
+		mtk_foe_entry_set_queue(eth, foe, queue);
+	else if (eth->qos_toggle == 1 || (ct_mark & MTK_QDMA_QUEUE_MASK) >= 9) {
+		u8 qos_ul_toggle;
+
+		if (eth->qos_toggle == 2)
+			qos_ul_toggle = ((ct_mark >> 16) & MTK_QDMA_QUEUE_MASK) >= 9 ? 1 : 0;
+		else
+			qos_ul_toggle = ((ct_mark >> 16) & MTK_QDMA_QUEUE_MASK) >= 1 ? 1 : 0;
+
+		if (qos_ul_toggle == 1) {
+			if (odev == eth->netdev[1])
+				mtk_foe_entry_set_queue(eth, foe, (ct_mark >> 16) & MTK_QDMA_QUEUE_MASK);
+			else
+				mtk_foe_entry_set_queue(eth, foe, ct_mark & MTK_QDMA_QUEUE_MASK);
+		} else
+			mtk_foe_entry_set_queue(eth, foe, ct_mark & MTK_QDMA_QUEUE_MASK);
 	}
-	mtk_foe_entry_set_queue(eth, foe, queue);
 
 out:
 	mtk_foe_entry_set_pse_port(eth, foe, pse_port);
@@ -480,8 +512,8 @@
 	if (data.pppoe.num == 1)
 		mtk_foe_entry_set_pppoe(eth, &foe, data.pppoe.sid);
 
-	err = mtk_flow_set_output_device(eth, &foe, odev, data.eth.h_dest,
-					 &wed_index);
+	err = mtk_flow_set_output_device(eth, &foe, idev, odev, f,
+					 data.eth.h_dest, &wed_index);
 	if (err)
 		return err;
 
--- a/include/net/flow_offload.h
+++ b/include/net/flow_offload.h
@@ -693,6 +693,7 @@
 	enum flow_cls_command command;
 	bool use_act_stats;
 	unsigned long cookie;
+	struct flow_offload *flow;
 	struct flow_rule *rule;
 	struct flow_stats stats;
 	u32 classid;
--- a/net/netfilter/nf_flow_table_offload.c
+++ b/net/netfilter/nf_flow_table_offload.c
@@ -884,11 +884,13 @@
 }
 
 static void nf_flow_offload_init(struct flow_cls_offload *cls_flow,
+				 struct flow_offload *flow,
 				 __be16 proto, int priority,
 				 enum flow_cls_command cmd,
 				 const struct flow_offload_tuple *tuple,
 				 struct netlink_ext_ack *extack)
 {
+	cls_flow->flow = flow;
 	cls_flow->common.protocol = proto;
 	cls_flow->common.prio = priority;
 	cls_flow->common.extack = extack;
@@ -910,7 +912,7 @@
 	__be16 proto = ETH_P_ALL;
 	int err, i = 0;
 
-	nf_flow_offload_init(&cls_flow, proto, priority, cmd,
+	nf_flow_offload_init(&cls_flow, flow, proto, priority, cmd,
 			     &flow->tuplehash[dir].tuple, &extack);
 	if (cmd == FLOW_CLS_REPLACE)
 		cls_flow.rule = flow_rule->rule;
